pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -507,104 +406,210 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
+			pr.Status.MarkResourceFailed(ReasonCouldntGetResource, wrapErr)
+			return
+		}
+		pipelineState, err := resources.ResolvePipelineRun(ctx,
+			*pr,
+			func(name string) (v1beta1.TaskInterface, error) {
+				return c.taskLister.Tasks(pr.Namespace).Get(name)
+			},
+			func(name string) (*v1beta1.TaskRun, error) {
+				return c.taskRunLister.TaskRuns(pr.Namespace).Get(name)
+			},
+			func(name string) (v1beta1.TaskInterface, error) {
+				return c.clusterTaskLister.Get(name)
+			},
+			func(name string) (*v1alpha1.Condition, error) {
+				return c.conditionLister.Conditions(pr.Namespace).Get(name)
+			},
+			pipelineSpec.Tasks, providedResources,
+		)
+		if err != nil {
+			// This Run has failed, so we need to mark it as failed and stop reconciling it
+			switch err := err.(type) {
+			case *resources.TaskNotFoundError:
+				wrapErr := fmt.Errorf("Pipeline %s can't be Run; it contains Tasks that don't exist: %s",
+					fmt.Sprintf("%s/%s", pipelineMeta.Namespace, pipelineMeta.Name), err)
+				pr.Status.MarkResourceFailed(ReasonCouldntGetTask, wrapErr)
+			case *resources.ConditionNotFoundError:
+				wrapErr := fmt.Errorf("PipelineRun %s can't be Run; it contains Conditions that don't exist:  %s",
+					fmt.Sprintf("%s/%s", pipelineMeta.Namespace, pr.Name), err)
+				pr.Status.MarkResourceFailed(ReasonCouldntGetCondition, wrapErr)
+			default:
+				wrapErr := fmt.Errorf("PipelineRun %s can't be Run; couldn't resolve all references: %s",
+					fmt.Sprintf("%s/%s", pipelineMeta.Namespace, pr.Name), err)
+				pr.Status.MarkResourceFailed(ReasonFailedValidation, wrapErr)
+			}
+		}
+		resolvedResultRefs := resources.ResolvePipelineResultRefs(pipelineState, pipelineSpec.Results)
+		pr.Status.PipelineResults = getPipelineRunResults(pipelineSpec, resolvedResultRefs)
+	}
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun, pipelineState resources.PipelineRunState, prDag *dag.Graph, pipelineSpec *v1beta1.PipelineSpec) error {
+
+	// We run "reconcile" once we know that the PipelineRun and its depedencies are valid and compatible.
+	// Results, since they are dynamic, may still cause validation issues.
+	//
+	// "reconcile" takes the following inputs:
+	// - ctx: the context
+	// - pr: the resource we are reconciling
+	// - pipelineState: the up-do-date statue of all TaskRuns and Conditions executed or running in the PipelineRun
+	// - prDag: the DAG calculated from the pipeline spec
+	// - pipelineSpec: the pipeline spec
+	//
+	// The structure of reconcile is:
+	// - Create any PVC that may be needed - when templates are used (only before the first task)
+	// - Find next Tasks from the DAG
+	// - Resolve results that may be input of Tasks to be started
+	// - Initialize (only when needed) the artifact storage
+	// - Loop through the tasks obtained from the DAG and run each of them
+	// - Update the overall PipelineRun status from pipelineState
+	// - Update the TaskRun status in PipelineRun status from the TaskRuns
```

Comment:
could this be a docstring? it seems like the kind of thing you'd have in a docstring